instrumentation/google.golang.org/grpc/otelgrpc/test/interceptor_test.go

```diff
@@ -583,39 +583,150 @@ func TestStreamClientInterceptorWithError(t *testing.T) {
+		spanCode: codes.Unset,
+		name:     "not.found",
+	},
+	{
+		grpcCode: grpc_codes.PermissionDenied,
+		spanCode: codes.Unset,
+		name:     "permission.denied",
+	},
+	{
+		grpcCode: grpc_codes.Unimplemented,
+		spanCode: codes.Error,
+		name:     "unimplemented",
+	},
+	{
+		grpcCode: grpc_codes.Internal,
+		spanCode: codes.Error,
+		name:     "internal",
+	},
+	{
+		grpcCode: grpc_codes.Unauthenticated,
+		spanCode: codes.Unset,
+		name:     "unauthenticated",
+	},
+}
+
+func UnaryServerInterceptor(t *testing.T) {
 	sr := tracetest.NewSpanRecorder()
 	tp := trace.NewTracerProvider(trace.WithSpanProcessor(sr))
 	usi := otelgrpc.UnaryServerInterceptor(otelgrpc.WithTracerProvider(tp))
-	deniedErr := status.Error(grpc_codes.PermissionDenied, "PERMISSION_DENIED_TEXT")
-	handler := func(_ context.Context, _ interface{}) (interface{}, error) {
-		return nil, deniedErr
-	}
-	_, err := usi(context.Background(), &grpc_testing.SimpleRequest{}, &grpc.UnaryServerInfo{}, handler)
-	require.Error(t, err)
-	assert.Equal(t, err, deniedErr)
+	for _, check := range serverChecks {
+		grpcErr := status.Error(check.grpcCode, check.grpcCode.String())
+		handler := func(_ context.Context, _ interface{}) (interface{}, error) {
+			return nil, grpcErr
+		}
+		_, err := usi(context.Background(), &grpc_testing.SimpleRequest{}, &grpc.UnaryServerInfo{FullMethod: check.name}, handler)
+		// the error should be nil only if the gRPC code is OK, otherwise it should be the same as the gRPC error
+		if check.grpcCode != grpc_codes.OK {
+			require.Error(t, err)
+			assert.Equal(t, err, grpcErr)
+		} else {
+			require.NoError(t, err)
+		}
 
-	span, ok := getSpanFromRecorder(sr, "")
-	if !ok {
-		t.Fatalf("failed to export error span")
-	}
-	assert.Equal(t, codes.Error, span.Status().Code)
-	assert.Contains(t, deniedErr.Error(), span.Status().Description)
-	var codeAttr attribute.KeyValue
-	for _, a := range span.Attributes() {
-		if a.Key == otelgrpc.GRPCStatusCodeKey {
-			codeAttr = a
-			break
+		span, ok := getSpanFromRecorder(sr, check.name)
+		if !assert.True(t, ok, "missing span %q", check.name) {
+			continue
+		}
+
+		// validate span status
+		assert.Equal(t, check.spanCode, span.Status().Code)
+		if span.Status().Code == codes.Error {
+			assert.Contains(t, grpcErr.Error(), span.Status().Description)
+		} else {
+			assert.Empty(t, span.Status().Description)
 		}
+
+		// validate span attributes
+		var codeAttr attribute.KeyValue
+		for _, a := range span.Attributes() {
+			if a.Key == otelgrpc.GRPCStatusCodeKey {
+				codeAttr = a
+				break
+			}
+		}
+		if assert.True(t, codeAttr.Valid(), "attributes contain gRPC status code") {
+			assert.Equal(t, attribute.Int64Value(int64(check.grpcCode)), codeAttr.Value)
+		}
+
+		// validate events and their attributes
+		assert.Len(t, span.Events(), 2)
+		assert.ElementsMatch(t, []attribute.KeyValue{
+			attribute.Key("message.type").String("SENT"),
+			attribute.Key("message.id").Int(1),
+		}, span.Events()[1].Attributes)
```

Comment:
Can we create some helper assertion methods to remove some duplication? I think it would also increase the readability.

Reply:
done